packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+import { isAbsolute, relative } from 'path';
+import { isValidPath } from '@graphql-tools/utils';
+import { normalizeInstanceOrArray, Types } from '@graphql-codegen/plugin-helpers';
+import isGlob from 'is-glob';
+import mm from 'micromatch';
+import { getLogger } from './logger.js';
+import logSymbols from 'log-symbols';
+import { CodegenContext } from '../config.js';
+
+function log(msg: string) {
+  // double spaces to inline the message with Listr
+  getLogger().info(`  ${msg}`);
+}
+
+type NegatedPattern = `!${string}`;
+
+/**
+ * Flatten a list of pattern sets to be a list of the original patterns that
+ * were used to create them (i.e., including both affirmative and negative).
+ *
+ * This can be used, for example, to find the "longest common prefix directory"
+ * by examining `mm.scan(pattern).base` for each `pattern`
+ */
+export const flattenPatternSets = (patternSets: PatternSet[]) => {
+  return patternSets.flatMap(patternSet => [
+    ...patternSet.watch.patterns,
+    ...patternSet.documents.patterns,
+    ...patternSet.schemas.patterns,
+  ]);
+};
+
+/**
+ * Create a rebuild trigger that follows the algorithm described here:
+ * https://github.com/dotansimha/graphql-code-generator/issues/9270#issuecomment-1496765045
+ *
+ * There is a flow chart diagram in that comment.
+ *
+ * Basically:
+ *
+ *  * "Global" patterns are defined at top level of config file, and "local"
+ *    patterns are defined for each output target
+ *  * Each pattern can have "watch", "documents", and "schemas"
+ *  * Watch patterns (global and local) always take precedence over documents and
+ *    schemas patterns, i.e. a watch negation always negates, and a watch match is
+ *    a match even if it would be negated by some pattern in documents or schemas
+ *  * The trigger returns true if any output target's local patterns result in
+ *    a match, after considering the precedence of any global and local negations
+ */
+export const makeShouldRebuild = ({
+  globalPatternSet,
+  localPatternSets,
+}: {
+  globalPatternSet: PatternSet;
+  localPatternSets: PatternSet[];
+}) => {
+  const localMatchers = localPatternSets.map(localPatternSet => {
+    return (path: string) => {
+      // Is path negated by any negating watch pattern?
+      if (
+        matchesAnyAffirmativePattern(
+          path,
+          invertNegatedPatterns([...globalPatternSet.watch.negated, ...localPatternSet.watch.negated])
+        )
+      ) {
+        // Short circut: negations in watch patterns take priority
+        return false;
+      }
+
+      // Does path match any affirmative watch pattern?
+      if (
+        matchesAnyAffirmativePattern(path, [
+          ...globalPatternSet.watch.affirmative,
+          ...localPatternSet.watch.affirmative,
+        ])
+      ) {
+        // Immediately return true: Watch pattern takes priority, even if documents or schema would negate it
+        return true;
+      }
```

Comment:
For watch patterns, we separate the patterns into `negated` and `affirmative` patterns, and we evaluate theme each separately. That is, we use `micromatch` only to evaluate affirmative patterns (or the inverted negations). This is necessary only for the `watch` rules because we use them to short circuit the algorithm.